Create a JSON file to store permission scopes required for ScubaGear

MichaelHicks-MSFT opened the pull request:
# <!-- Use the title to describe PR changes in the imperative mood --> #
  <!-- Remember this title will end up as the merge commit subject -->

## 🗣 Description ##
Permissions file containing current SCuBAGear permissions mapping.
<!-- Describe the "what" of your changes in detail. -->
<!-- To avoid scope creep, limit changes to a single goal. -->

## 💭 Motivation and context ##
Added permissions file to allow centralized management of all SCuBAGear Microsoft Graph permissions.

https://github.com/cisagov/ScubaGear/issues/1391 is a follow up to this issue, which aims to modify the Connection module to read from the new ScubaPermissions.json file instead of hard coded values.
<!-- Why is this change required? -->
<!-- What problem does this change solve? How did you solve it? -->
<!-- Mention any related issue(s) here using appropriate keywords such -->
<!-- as "closes" or "resolves" to auto-close them on merge. -->
Closes #1356 

## 🧪 Testing ##
Testing was conducted across commercial and gcchigh tenants. Created PowerShell code that imported the JSON file and made connections to Microsoft Graph utilizing the permission scopes defined under Permission within the ScubaGearGraphScopes section of the file.
<!-- How did you test your changes? How could someone else test this PR? -->
<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->

<!--
## 📷 Screenshots (if appropriate) ##

Uncomment this section if a screenshot is needed.

-->

## ✅ Pre-approval checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs may have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

- [x] This PR has an informative and human-readable title.
- [x] PR targets the correct parent branch (e.g., main or release-name) for merge.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] Changes are sized such that they do not touch excessive number of files.
- [x] *All* future TODOs are captured in issues, which are referenced in code comments.
- [x] These code changes follow the ScubaGear [content style guide](https://github.com/cisagov/ScubaGear/blob/main/CONTENTSTYLEGUIDE.md).
- [x] Related issues these changes resolve are linked preferably via [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).
- [x] All relevant type-of-change labels added.
- [x] All relevant project fields are set.
- [x] All relevant repo and/or project documentation updated to reflect these changes.
- [x] Unit tests added/updated to cover PowerShell and Rego changes.
- [x] Functional tests added/updated to cover PowerShell and Rego changes.
- [x] All relevant functional tests passed.
- [x] All automated checks (e.g., linting, static analysis, unit/smoke tests) passed.

## ✅ Pre-merge checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- These boxes should remain unchecked until the pull request has been -->
<!-- approved. -->

- [ ] PR passed smoke test check.
- [ ] Feature branch has been rebased against changes from parent branch, as needed

  Use `Rebase branch` button below or use [this](https://www.digitalocean.com/community/tutorials/how-to-rebase-and-update-a-pull-request) reference to rebase from the command line.
- [ ] Resolved all merge conflicts on branch
- [ ] Notified merge coordinator that PR is ready for merge via comment mention

## ✅ Post-merge checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- These boxes should remain unchecked until the pull request has been -->
<!-- approved. This section is for the merge coordinator to complete. -->
- [ ] Feature branch deleted after merge to clean up repository.
- [ ] Verified that all checks pass on parent branch (e.g., main or release-name) after merge.

tkol2022 reviewed (changes requested):
- Can you move the file from the /Connection folder to a new /Permissions folder instead?
- I noticed you removed the ScubaGearSPPermissions node. In the future ScubaGear will have a utility script that helps users register a service principal to run the tool. Is the thought that our new utility script can just filter the GraphCmdLetPermissions node for entries that are of runtype = application and just use that list to create the needed permissions for the service principal in Entra Id?
- We need to add the following REST APIs as entries to the GraphCmdLetPermissions node along with their associated required permissions. Entra Id now includes direct calls to REST APIs without Cmdlets and therefore we need to ensure those permissions are represented.
  - /beta/roleManagement/directory/roleEligibilityScheduleInstances
  - /beta/roleManagement/directory/roleAssignmentScheduleInstances
  - /beta/identityGovernance/privilegedAccess/group/eligibilityScheduleInstances
  - /beta/privilegedAccess/aadGroups/resources

mitchelbaker-cisa commented:
Updated leastPermissions and poshModule keys to array type.

mitchelbaker-cisa commented on PowerShell/ScubaGear/Modules/Permissions/ScubaGearPermissions.json:
```diff
@@ -0,0 +1,616 @@
+
+    {
+        "moduleCmdlet": "Connect-ExchangeOnline",
+        "apiResource": {
```
> This is a multi-valued field that sometimes is a string and sometimes a dictionary. As such, recommend always representing as a dictionary. For cases where the API resource is the same across all environments, perhaps use key of "all" or "any"? Either that or replicate URI across all keys in supportedEnv for that entry. The bit of duplication seems reasonable to make it easier to parse.

@MichaelHicks-MSFT could you assist with updating all the "apiResource" keys to be of type dictionary? Please see Addam's comment above. I think for API resources which are the same across all environments, we replicate the URI across all keys. If an environment is not supported for the given resource, set it to null. 

```
{
  "moduleCmdlet": "Connect-PnPOnline",
  "apiResource": {
    "commercial": "https://{domain}-admin.sharepoint.com",
    "gcc": "https://{domain}-admin.sharepoint.com",
    "gcchigh": "https://{domain}-admin.sharepoint.us",
    "dod": "https://{domain}-admin.sharepoint.mil.us"
  },
  ...
}
```

mitchelbaker-cisa commented on PowerShell/ScubaGear/Modules/Permissions/ScubaGearPermissions.json:
```diff
@@ -0,0 +1,674 @@
+            "Directory.ReadWrite.All",
+            "Directory.Read.All"
+        ],
+        "permissionType": "Application",
```
`permissionType` to be removed since its no longer used.

MichaelHicks-MSFT commented:
Updated permissions file to reflect discussion last week. Unified keys across objects, renamed rolePermissions to spRolePermissions. Created singular objects for connect cmdlets to break out apiResource based on the environment.

mitchelbaker-cisa commented on PowerShell/ScubaGear/Modules/Permissions/ScubaGearPermissions.json:
```diff
@@ -0,0 +1,674 @@
+            "dod": "https://{domain}-admin.sharepoint.mil.us"
+        },
+        "poshModule": ["Microsoft.Online.SharePoint.PowerShell"],
+        "leastPermissions": ["Sites.FullControl.All"],
```
A new property was added to indicate noninteractive permissions, `"spRolePermissions": []`. @buidav does that resolve your comment? Or are additional changes needed.

tkol2022 commented:
### Testing Instructions for Interactive authentication
The purpose of this set of tests is to exercise the new pieces of code with an interactive login. There will be separate instructions for testing with service principal.

#### Preparation
Download the main branch and run ScubaGear to create a local scuba results json file which will be used later in a comparison. Run with aad, defender and sharepoint. Include the same products when you run the new code in the steps below.

Download https://github.com/MichaelHicks-MSFT/ScubaGear/tree/main to a local folder /PermissionsJSON 

#### Test to ensure Get-ScubaGearPermissions sets up up the correct permissions for ScubaGear to run interactively
Go to Microsoft Graph Powershell in the Entra Id > Enterprise Applications and click on Permissions
Delete the application from the tenant (this way the permissions will be cleaned out and then recreated the next time that ScubaGear is run)
Go to the PermissionsJSON folder and run ScubaGear interactively. Consent to the permissions for the organization (this is important because it allows others on the team to be able to run ScubaGear interactively).
This should create a Scuba results file

Compare this file against the Scuba results file created from the main branch and ensure that there are no differences that would indicate that the new Powershell code changes had any material effect on the items below. Ignore timestamp and ReportUUID differences.

- The provider JSON that gets created
- The results of the Rego tests

If there are differences between the two files, this indicates that the introduction of the Get-ScubaGearPermissions cmdlet may have introduced a bug.

#### Test to ensure Get-ScubaGearPermissions does not generate any permissions errors (e.g. 401)
Examine the Powershell window where the PermissionsJSON branch was executed and note any permissions errors (e.g. "unauthorized")

#### Test to determine if there are any differences in the permissions created in the tenant that require a review
Go to the Entra Id > Enterprise Applications and open the permissions for the Microsoft Graph Powershell app. These are the permissions created by running the PermissionsJSON with the new Get-ScubaGearPermissions cmdlet in the Connect-Tenant code.
Compare the permissions to another tenant for the same app Microsoft Graph Powershell and note any differences

tkol2022 commented:
### Test Results for Interactive authentication - E5 tenant

#### Test 1 - Ensure Get-ScubaGearPermissions sets up up the correct permissions for ScubaGear to run interactively
Compared the configuration settings export and policy check results. No difference between the main branch versus this one. Pass.

#### Test 2 - Ensure Get-ScubaGearPermissions does not generate any permissions errors (e.g. 401)
No permissions errors observed in the console. Pass.

#### Test 3 -Determine if there are any differences in the permissions created in the tenant that require a review
Some new permissions that were not there before were requested for the MS Graph Powershell (Microsoft Graph Command Line Tools) service principal. After discussion with the MS team, this is because each of the cmdlets documented in the JSON permissions include the least privileged permission required to execute the respective cmdlet. Pass.

tkol2022 commented:
### Test Results for Interactive authentication - G3 tenant

#### Test 1 - Ensure Get-ScubaGearPermissions sets up up the correct permissions for ScubaGear to run interactively
Pass.

#### Test 2 - Ensure Get-ScubaGearPermissions does not generate any permissions errors (e.g. 401)
Pass.

#### Test 3 -Determine if there are any differences in the permissions created in the tenant that require a review
Pass.

tkol2022 commented:
### Test Results for Interactive authentication - GCC High tenant

#### Test 1 - Ensure Get-ScubaGearPermissions sets up up the correct permissions for ScubaGear to run interactively
Pass.

#### Test 2 - Ensure Get-ScubaGearPermissions does not generate any permissions errors (e.g. 401)
Pass.

#### Test 3 -Determine if there are any differences in the permissions created in the tenant that require a review
Pass.

tkol2022 commented:
Should the cmdlet Get-ScubaGearEntraRedundantPermissions be renamed to Get-ScubaGearEntraMinimumPermissions? hopefully I understood the intent of the function.

![image](https://github.com/user-attachments/assets/c5c23b3c-3c49-4ed5-a233-dad72cd73057)

tkol2022 commented:
### Test Results for Non-Interactive authentication - GCC High, E5 and G3 tenants
The tests were performed on all tenant types listed above. I created a brand new application to test with in each tenant and gave it the permissions listed by the Get-ScubaGearEntraRedundantPermissions cmdlet.

#### Test 1 - Ensure that Connect-Tenant in ScubaGear works correctly with non interactive auth and ScubaGear produces the expected output json and policy results
Pass.

#### Test 2 - Ensure Get-ScubaGearPermissions does not generate any permissions errors (e.g. 401)
Pass.

tkol2022 commented:
### Testing changes to Invoke-GraphDirectly
The AAD export provider function Invoke-GraphDirectly was tested indirectly by executing ScubaGear against Entra Id in three tenant types. No issues observed.

tkol2022 commented:
Tested parameter validation of Get-ScubaGearPermissions and behaves as expected.

![image](https://github.com/user-attachments/assets/c4048d5b-f46c-436c-8f7a-dcc845446ae0)

tkol2022 commented:
I was using ScubaGearPermissions.json as a reference for another issue that I'm working on and I noticed we are missing Get-MgBetaDirectoryObject from the file.

MichaelHicks-MSFT commented:
> I was using ScubaGearPermissions.json as a reference for another issue that I'm working on and I noticed we are missing Get-MgBetaDirectoryObject from the file.

I will add that to the file.

tkol2022 commented:
This PR is being[ recreated with a new branch directly in the ScubaGear repo](https://github.com/cisagov/ScubaGear/pull/1597).
